File: devboard-frontend/src/services/taskService.js
import api from './api'

export const taskService = {
  // Get all tasks
  getAllTasks: async () => {
    try {
      const response = await api.get('/tasks')
      return response.data
    } catch (error) {
      throw new Error(`Failed to fetch tasks: ${error.message}`)
    }
  },

  // Get tasks with filters
  getTasksWithFilters: async (filters = {}) => {
    try {
      const params = new URLSearchParams()
      
      if (filters.assigneeId) params.append('assigneeId', filters.assigneeId)
      if (filters.priority) params.append('priority', filters.priority)
      if (filters.status) params.append('status', filters.status)
      if (filters.search) params.append('search', filters.search)
      if (filters.creatorId) params.append('creatorId', filters.creatorId)
      
      const url = params.toString() ? `/tasks?${params.toString()}` : '/tasks'
      const response = await api.get(url)
      return response.data
    } catch (error) {
      throw new Error(`Failed to fetch filtered tasks: ${error.message}`)
    }
  },

  // Get task by ID
  getTaskById: async (id) => {
    try {
      const response = await api.get(`/tasks/${id}`)
      return response.data
    } catch (error) {
      throw new Error(`Failed to fetch task ${id}: ${error.message}`)
    }
  },

  // Create new task
  createTask: async (task) => {
    try {
      const response = await api.post('/tasks', task)
      return response.data
    } catch (error) {
      throw new Error(`Failed to create task: ${error.message}`)
    }
  },

  // Update task
  updateTask: async (id, task) => {
    try {
      const response = await api.put(`/tasks/${id}`, task)
      return response.data
    } catch (error) {
      throw new Error(`Failed to update task ${id}: ${error.message}`)
    }
  },

  // Delete task
  deleteTask: async (id) => {
    try {
      await api.delete(`/tasks/${id}`)
      return true
    } catch (error) {
      throw new Error(`Failed to delete task ${id}: ${error.message}`)
    }
  },

  // Get tasks by status
  getTasksByStatus: async (status) => {
    try {
      const response = await api.get(`/tasks/status/${status}`)
      return response.data
    } catch (error) {
      throw new Error(`Failed to fetch tasks with status ${status}: ${error.message}`)
    }
  },

  // Get tasks by priority
  getTasksByPriority: async (priority) => {
    try {
      const response = await api.get(`/tasks/priority/${priority}`)
      return response.data
    } catch (error) {
      throw new Error(`Failed to fetch tasks with priority ${priority}: ${error.message}`)
    }
  },

  // Get task detail with comments
  getTaskDetail: async (id) => {
    try {
      const response = await api.get(`/tasks/${id}/detail`)
      return response.data
    } catch (error) {
      throw new Error(`Failed to fetch task detail ${id}: ${error.message}`)
    }
  },

  // Add comment to task
  addComment: async (taskId, comment) => {
    try {
      const response = await api.post(`/tasks/${taskId}/comments`, comment)
      return response.data
    } catch (error) {
      throw new Error(`Failed to add comment to task ${taskId}: ${error.message}`)
    }
  },

  // Get comments for task
  getTaskComments: async (taskId) => {
    try {
      const response = await api.get(`/tasks/${taskId}/comments`)
      return response.data
    } catch (error) {
      throw new Error(`Failed to fetch comments for task ${taskId}: ${error.message}`)
    }
  },

  // Delete comment
  deleteComment: async (commentId) => {
    try {
      await api.delete(`/comments/${commentId}`)
      return true
    } catch (error) {
      throw new Error(`Failed to delete comment ${commentId}: ${error.message}`)
    }
  },

  // Test API connection
  testConnection: async () => {
    try {
      const response = await api.get('/hello')
      return response.data
    } catch (error) {
      throw new Error(`Failed to connect to API: ${error.message}`)
    }
  }
}

export default taskService

File: devboard-backend/src/main/java/com/example/devboard/service/TaskService.java
package com.example.devboard.service;

import com.example.devboard.common.ErrorCode;
import com.example.devboard.dto.TaskCreateRequest;
import com.example.devboard.dto.TaskResponse;
import com.example.devboard.dto.TaskUpdateRequest;
import com.example.devboard.dto.TaskDetailResponse;
import com.example.devboard.dto.CommentResponse;
import com.example.devboard.entity.Task;
import com.example.devboard.entity.User;
import com.example.devboard.exception.BusinessException;
import com.example.devboard.repository.TaskRepository;
import com.example.devboard.repository.UserRepository;
import com.example.devboard.repository.CommentRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Transactional
@Slf4j
public class TaskService {
    
    private final TaskRepository taskRepository;
    private final UserRepository userRepository;
    private final CommentRepository commentRepository;
    
    public List<TaskResponse> getAllTasks() {
        return taskRepository.findAll().stream()
                .map(this::convertToResponse)
                .collect(Collectors.toList());
    }
    
    public List<TaskResponse> getAllTasksWithFilters(Long assigneeId, String priority, String status, String search, Long creatorId) {
        return taskRepository.findAll().stream()
                .filter(task -> {
                    // Filter by assignee - only filter if assigneeId is specified
                    if (assigneeId != null) {
                        // If task has no assignee, exclude it when looking for specific assignee
                        if (task.getAssignee() == null) {
                            return false;
                        }
                        // If task has assignee but doesn't match the filter, exclude it
                        if (!task.getAssignee().getId().equals(assigneeId)) {
                            return false;
                        }
                    }
                    
                    // Filter by creator - only filter if creatorId is specified
                    if (creatorId != null) {
                        if (task.getCreator() == null || !task.getCreator().getId().equals(creatorId)) {
                            return false;
                        }
                    }
                    
                    // Filter by priority - only filter if priority is specified
                    if (priority != null && !priority.isEmpty()) {
                        try {
                            Task.TaskPriority taskPriority = Task.TaskPriority.valueOf(priority.toUpperCase());
                            if (!task.getPriority().equals(taskPriority)) {
                                return false;
                            }
                        } catch (IllegalArgumentException e) {
                            log.warn("Invalid priority filter: {}", priority);
                            return false;
                        }
                    }
                    
                    // Filter by status - only filter if status is specified
                    if (status != null && !status.isEmpty()) {
                        try {
                            Task.TaskStatus taskStatus = Task.TaskStatus.valueOf(status.toUpperCase());
                            if (!task.getStatus().equals(taskStatus)) {
                                return false;
                            }
                        } catch (IllegalArgumentException e) {
                            log.warn("Invalid status filter: {}", status);
                            return false;
                        }
                    }
                    
                    // Search in title and description - only filter if search is specified
                    if (search != null && !search.trim().isEmpty()) {
                        String searchLower = search.toLowerCase().trim();
                        boolean titleMatch = task.getTitle() != null && 
                                           task.getTitle().toLowerCase().contains(searchLower);
                        boolean descMatch = task.getDescription() != null && 
                                          task.getDescription().toLowerCase().contains(searchLower);
                        if (!titleMatch && !descMatch) {
                            return false;
                        }
                    }
                    
                    return true;
                })
                .map(this::convertToResponse)
                .collect(Collectors.toList());
    }
    
    public TaskResponse getTaskById(Long id) {
        Task task = taskRepository.findById(id)
                .orElseThrow(() -> new BusinessException(ErrorCode.TASK_NOT_FOUND, "Task not found with id: " + id));
        return convertToResponse(task);
    }
    
    public TaskDetailResponse getTaskDetail(Long id) {
        Task task = taskRepository.findById(id)
                .orElseThrow(() -> {
                    log.error("Task not found with id: {}", id);
                    return new BusinessException(ErrorCode.TASK_NOT_FOUND, "Task not found with id: " + id);
                });
        
        TaskResponse taskResponse = convertToResponse(task);
        List<CommentResponse> comments = commentRepository.findByTaskIdOrderByCreatedAtDesc(id)
                .stream()
                .map(comment -> {
                    CommentResponse.UserSummary userSummary = CommentResponse.UserSummary.builder()
                            .id(comment.getUser().getId())
                            .username(comment.getUser().getUsername())
                            .nickname(comment.getUser().getNickname())
                            .avatar(comment.getUser().getAvatar())
                            .build();
                    
                    return CommentResponse.builder()
                            .id(comment.getId())
                            .content(comment.getContent())
                            .user(userSummary)
                            .createdAt(comment.getCreatedAt())
                            .updatedAt(comment.getUpdatedAt())
                            .build();
                })
                .collect(Collectors.toList());
        
        return TaskDetailResponse.fromTaskResponse(taskResponse, comments);
    }
    
    public TaskResponse createTask(TaskCreateRequest request, Long creatorId) {
        User creator = userRepository.findById(creatorId)
                .orElseThrow(() -> new BusinessException(ErrorCode.USER_NOT_FOUND, "Creator not found"));
        
        Task task = Task.builder()
                .title(request.getTitle())
                .description(request.getDescription())
                .status(Task.TaskStatus.valueOf(request.getStatus()))
                .priority(Task.TaskPriority.valueOf(request.getPriority()))
                .creator(creator)
                .build();
        
        // Set assignee if provided
        if (request.getAssigneeId() != null) {
            User assignee = userRepository.findById(request.getAssigneeId())
                    .orElseThrow(() -> new BusinessException(ErrorCode.USER_NOT_FOUND, "Assignee not found"));
            task.setAssignee(assignee);
        }
        
        Task savedTask = taskRepository.save(task);
        return convertToResponse(savedTask);
    }
    
    public TaskResponse updateTask(Long id, TaskUpdateRequest request, Long userId) {
        Task task = taskRepository.findById(id)
                .orElseThrow(() -> {
                    log.error("Task not found with id: {}", id);
                    return new BusinessException(ErrorCode.TASK_NOT_FOUND, "Task not found with id: " + id);
                });
        
        log.info("User {} updating task {} - '{}'", userId, id, task.getTitle());
        
        // New permission logic: Any authenticated user can update any task
        
        // Update fields if provided
        if (request.getTitle() != null) {
            task.setTitle(request.getTitle());
        }
        if (request.getDescription() != null) {
            task.setDescription(request.getDescription());
        }
        if (request.getStatus() != null) {
            task.setStatus(Task.TaskStatus.valueOf(request.getStatus()));
        }
        if (request.getPriority() != null) {
            task.setPriority(Task.TaskPriority.valueOf(request.getPriority()));
        }
        if (request.getAssigneeId() != null) {
            User assignee = userRepository.findById(request.getAssigneeId())
                    .orElseThrow(() -> new BusinessException(ErrorCode.USER_NOT_FOUND, "Assignee not found"));
            task.setAssignee(assignee);
        }
        
        Task updatedTask = taskRepository.save(task);
        return convertToResponse(updatedTask);
    }
    
    public void deleteTask(Long id, Long userId) {
        Task task = taskRepository.findById(id)
                .orElseThrow(() -> {
                    log.error("Task not found with id: {}", id);
                    return new BusinessException(ErrorCode.TASK_NOT_FOUND, "Task not found with id: " + id);
                });
        
        // Get user to check if admin
        User user = userRepository.findById(userId)
                .orElseThrow(() -> {
                    log.error("User not found with id: {}", userId);
                    return new BusinessException(ErrorCode.USER_NOT_FOUND, "User not found");
                });
        
        // Only creator or admin can delete
        boolean isCreator = task.getCreator() != null && task.getCreator().getId().equals(userId);
        boolean isAdmin = user.getRole() == User.UserRole.ADMIN;
        
        if (!isCreator && !isAdmin) {
            log.warn("User {} attempted to delete task {} - permission denied. Creator: {}, User role: {}", 
                    userId, id, 
                    task.getCreator() != null ? task.getCreator().getId() : "null", 
                    user.getRole());
            throw new BusinessException(ErrorCode.TASK_ACCESS_DENIED, "Only the creator or admin can delete this task");
        }
        
        log.info("User {} deleting task {} - '{}'", userId, id, task.getTitle());
        taskRepository.delete(task);
    }
    
    public List<TaskResponse> getTasksByStatus(String status) {
        Task.TaskStatus taskStatus = Task.TaskStatus.valueOf(status);
        return taskRepository.findByStatus(taskStatus).stream()
                .map(this::convertToResponse)
                .collect(Collectors.toList());
    }
    
    public List<TaskResponse> getMyTasks(Long userId) {
        List<Task> tasks = taskRepository.findAll().stream()
                .filter(task -> (task.getCreator() != null && task.getCreator().getId().equals(userId)) || 
                               (task.getAssignee() != null && task.getAssignee().getId().equals(userId)))
                .collect(Collectors.toList());
        
        return tasks.stream()
                .map(this::convertToResponse)
                .collect(Collectors.toList());
    }
    
    private TaskResponse convertToResponse(Task task) {
        TaskResponse.UserSummary creator = null;
        if (task.getCreator() != null) {
            creator = TaskResponse.UserSummary.builder()
                    .id(task.getCreator().getId())
                    .username(task.getCreator().getUsername())
                    .nickname(task.getCreator().getNickname())
                    .avatar(task.getCreator().getAvatar())
                    .build();
        }
        
        TaskResponse.UserSummary assignee = null;
        if (task.getAssignee() != null) {
            assignee = TaskResponse.UserSummary.builder()
                    .id(task.getAssignee().getId())
                    .username(task.getAssignee().getUsername())
                    .nickname(task.getAssignee().getNickname())
                    .avatar(task.getAssignee().getAvatar())
                    .build();
        }
        
        return TaskResponse.builder()
                .id(task.getId())
                .title(task.getTitle())
                .description(task.getDescription())
                .status(task.getStatus().name())
                .priority(task.getPriority().name())
                .creator(creator)
                .assignee(assignee)
                .createdAt(task.getCreatedAt())
                .updatedAt(task.getUpdatedAt())
                .commentCount(commentRepository.countByTaskId(task.getId()))
                .build();
    }
    
    // Permission checking method for @PreAuthorize
    public boolean isTaskCreator(Long taskId, String username) {
        return taskRepository.findById(taskId)
                .map(task -> task.getCreator().getUsername().equals(username))
                .orElse(false);
    }
    
    // Admin-only task deletion
    public void adminDeleteTask(Long taskId) {
        Task task = taskRepository.findById(taskId)
                .orElseThrow(() -> new BusinessException(ErrorCode.TASK_NOT_FOUND, "Task not found"));
        
        log.info("Admin deleting task: {} ({})", task.getId(), task.getTitle());
        taskRepository.delete(task);
    }
}